### DISCUSSION TITLE
Lesson10: why we write vm.prank in setUp function before ourToken.transfer?

### DISCUSSION THREAD

**Discussion by codeplay22** [Cyfrin/foundry-full-course-cu#1916] [Q&A] [answered]
When i don't type ```vm.prank(msg.sender)``` in setUp function this error comes, but when i check the balance of ```address(this) 
it's 79228162414264337593543950335``` which is much higher than ```STARTING_BALANCE which is 100 ether``` and if i also check the balance of  ``` transfer function in ERC20 Which is coming out to be 79228162414264337593543950335 only``` still it's giving me ERC20InsufficientBalance (you can see it in the image down below). Why this error is coming? it should not come because owner have balance of 79228162414264337593543950335 which is much higher than 100 ethers, so it should tranfer it??

![image](https://github.com/Cyfrin/foundry-full-course-cu/assets/116997721/ec14ce39-02ef-493f-9fd4-6522abc97474)

This is my setUp function 
```    function setUp() public {
        deployer = new DeployOurToken();
        ourToken = deployer.run();
        console.log("address of address(this) in OurTokenTest.t.sol is", address(this));
        console.log("msg.sender balance is", msg.sender.balance);
        console.log("address(this).balance", address(this).balance - (STARTING_BALANCE));
        ourToken.transfer(bob, STARTING_BALANCE);
    }
 ```

**Comment by EngrPips** [comment]
Hello @codeplay22, Please show your full `setup` function.

**Reply by codeplay22** [reply]
hi @EngrPips 
this is it
```     
function setUp() public {
        deployer = new DeployOurToken();
        ourToken = deployer.run();
        console.log("address of address(this) in OurTokenTest.t.sol is", address(this));
        // vm.prank(address(msg.sender));
        console.log("msg.sender balance is", msg.sender.balance);
        console.log("address(this).balance", address(this).balance - (STARTING_BALANCE));
        // console.log("balanceof(address(this))", balanceof(address(this)));
        ourToken.transfer(bob, STARTING_BALANCE);
    }
```

**Reply by EngrPips** [reply]
Good. Please show `DeployOurToken` and `ourToken`  files.

**Reply by codeplay22** [reply]
DeployOurToken.s.sol
```
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "../src/OurToken.sol";
import {console} from "forge-std/Test.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;

    function run() external returns (OurToken) {
        vm.startBroadcast();
        OurToken ot = new OurToken(INITIAL_SUPPLY);
        // console.log("address(this) in DeployOurToken.s.sol is",address(this));
        // console.log("address of ot is",address(ot));
        // console.log("msg.sender is in DeployOurToken.s.sol is",msg.sender);
        vm.stopBroadcast();
        return ot;
    }
}
```

ManualToken.sol
```
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;

contract ManualToken {

    mapping (address => uint256) private s_balance;

    function name() public pure returns (string memory){
        return "Manual Token";
    }

    function balanceOf(address _owner) public view  returns (uint256 balance){
        return s_balance[_owner];
    }

    function decimals() public pure returns (uint8){
        return 18;
    }

    function totalSupply() public pure returns (uint256) {
        return 100 ether;//100000000000
    }

    function transfer(address _to, uint256 _amount) public {
        uint256 previousBalances = balanceOf(msg.sender) + balanceOf(_to);
        s_balance[msg.sender] -= _amount;
        s_balance[_to] += _amount;
        require(balanceOf(msg.sender) + balanceOf(_to) == previousBalances); 
    }





}
```

OurToken.sol
```
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import {console} from "forge-std/Test.sol";
contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply);
        console.log("address(this) inside OurToken.sol is",address(this));
        console.log("msg.sender in OurToken.sol is",msg.sender);
    }    

}
```

**Reply by EngrPips** [reply, marked as answer]
```js
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;

import {Script} from "forge-std/Script.sol";
import {OurToken} from "../src/OurToken.sol";
import {console} from "forge-std/Test.sol";

contract DeployOurToken is Script {
    uint256 public constant INITIAL_SUPPLY = 1000 ether;

    function run() external returns (OurToken) {
        vm.startBroadcast(); <@ here
        OurToken ot = new OurToken(INITIAL_SUPPLY);
        // console.log("address(this) in DeployOurToken.s.sol is",address(this));
        // console.log("address of ot is",address(ot));
        // console.log("msg.sender is in DeployOurToken.s.sol is",msg.sender);
        vm.stopBroadcast();
        return ot;
    }
}
```
So, in this deployer script, you used `vm.startBroadcast()` and did not pass in a privateKey, so that will make foundry use its default privateKey. Take note of this, as you will need to understand the next code.

```js
// SPDX-License-Identifier: SEE LICENSE IN LICENSE
pragma solidity ^0.8.18;
import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
import {console} from "forge-std/Test.sol";
contract OurToken is ERC20{

    constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply); <@ here
        console.log("address(this) inside OurToken.sol is",address(this));
        console.log("msg.sender in OurToken.sol is",msg.sender);
    }    

}
```
So in your `OurToken` contract, you mint `initialSupply` amount of `OurToken` to the `msg.sender`, which will be the default foundry sender because you did not pass in any privateKey when you used `vm.startBroadcast()` in your deployment script. 

```js
function setUp() public {
        deployer = new DeployOurToken();
        ourToken = deployer.run();
        console.log("address of address(this) in OurTokenTest.t.sol is", address(this));
        // vm.prank(address(msg.sender));
        console.log("msg.sender balance is", msg.sender.balance);
        console.log("address(this).balance", address(this).balance - (STARTING_BALANCE));
        // console.log("balanceof(address(this))", balanceof(address(this)));
        ourToken.transfer(bob, STARTING_BALANCE); <@ here
    }
```
So here you are trying to transfer token from the test script to `bob,` but the test did not have any balance on `OurToken` because the `initialSupply` was minted to foundry default sender. So your test reverts with `ERC20InsufficientBalance.` If anything is confusing, please ask question

**Reply by codeplay22** [reply]
@EngrPips 
i did'nt understand the last part which is this
 ``` the test did not have any balance on OurToken because the initialSupply was minted to foundry default sender. ```
 
 can you explain those lines more deeply and easily?, i think i didnt understand that english.
 
 did you mean that we didnt use test to mint ```initialSupply``` inside the constructor?

**Reply by EngrPips** [reply]
I mean, your token will mint to the `foundry default sender` address because of the `vm.startBroadcast()`, which you didn't pass any private key to 

```js
constructor(uint256 initialSupply) ERC20("OurToken","OT"){
        _mint(msg.sender,initialSupply); <@ here
        console.log("address(this) inside OurToken.sol is",address(this));
        console.log("msg.sender in OurToken.sol is",msg.sender);
    }    
```
This line `_mint(msg.sender,initialSupply); <@ here` mints to the `foundry default sender,` but in your test file, it is the test address that is trying to send token to `bob.`  The test address does not hold any token, so the test reverted with. `ERC20InsufficientBalance`. If anything is confusing, please ask question

**Reply by codeplay22** [reply]
okay,
so how can i check token of msg.sender then?

**Reply by EngrPips** [reply]
How do you mean check token of `msg.sender`? can you please explain in a more simple way?

**Reply by codeplay22** [reply]
you said this ``` The test address does not hold any token, so the test reverted with  ERC20InsufficientBalance```
so how can we check token of a particular address in this case it  is msg.sender. so how can we check token of msg.sender then?

**Comment by EngrPips** [comment]
```js
deployer = new DeployOurToken();
        ourToken = deployer.run();
        console.log("address of address(this) in OurTokenTest.t.sol is", address(this));
        console.log("msg.sender balance is", msg.sender.balance);
        console.log("address(this).balance", address(this).balance - (STARTING_BALANCE));
        console.log("The balance of this contract is: , ourToken.balanceOf(address(this)));
        ourToken.transfer(bob, STARTING_BALANCE);
    }
```
Run this code, and it should print the `ourToken`  balance of your test contract.

**Reply by EngrPips** [reply]
Please ask questions if anything is still confusing.

**Reply by codeplay22** [reply]
@EngrPips thanks u for your help
so basically ```it's not mapped to msg.sender's address ``` that's why it's giving me insufficient balance because i didn't mint token with ```address(this)``` address so that's why its giving balance of 0 because test's address(address(this)) not mapped to any token(balance). 
your explaination was a bit confusing to me but still your code helped me to thinking about it and got to this. 
So thanks a lot!!

**Reply by EngrPips** [reply]
I am very sorry that I explained things poorly. I tried to do the best I could. I am glad you at least figured it out. Keep crushing it, buddy.

**Reply by codeplay22** [reply]
no worries
Thanks a lot! you too!

**Reply by EngrPips** [reply]
I appreciate you, Man.
